### PR TITLE
Make "server socket not listening" warning logs less noisy

### LINKED ISSUE #73114: nightly: excessive The server socket on ... is not yet listening
### 🐛 Describe the bug

Switching from pt-1.10 to nightly on HPC I'm now getting hundreds of:

```
[W socket.cpp:701] The server socket on [jean-zay-iam01-ib0]:55035 is not yet listening (errno: 111 - Connection refused), will retry
```
every time I start a distributed training. Several times the printing won't stop and I had to kill the job.

I don't propose to remove it, as it's important, but perhaps it doesn't have to spam the logs? Log it once a minute maybe?

@cbalioglu

### Versions

nightly 2022-02-16

### PR DESCRIPTION
**Pull request by cbalioglu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #73149
* #73167
* #73166
* #73144

This PR improves the handling of the "server socket not yet listening" warning log in c10d `socket`. Instead of outputting it after every failed attempt (meaning every second), it is now written every 20 seconds. Note though that if the log level is set to `INFO`, we keep writing a detailed message every second as before with additional `errno` information.

With log level set to `WARN` the output looks like:
```
[W socket.cpp:598] [c10d] No socket on (127.0.0.1, 29501) is listening yet, will retry.
[W socket.cpp:598] [c10d] No socket on (127.0.0.1, 29501) is listening yet, will retry.
...
[E socket.cpp:726] [c10d] The client socket has timed out after 300s while trying to connect to (127.0.0.1, 29501).
```

With log level set to `INFO` (a.k.a. verbose or debug level) the output looks like:
```
[I socket.cpp:515] [c10d] The client socket will attempt to connect to an IPv6 address of (127.0.0.1, 29501).
[I socket.cpp:582] [c10d] The client socket is attempting to connect to [localhost]:29501.
[I socket.cpp:643] [c10d] The server socket on [localhost]:29501 is not yet listening (errno: 111 - Connection refused), will retry.
[W socket.cpp:598] [c10d] No socket on (127.0.0.1, 29501) is listening yet, will retry.
[I socket.cpp:582] [c10d] The client socket is attempting to connect to [localhost]:29501.
[I socket.cpp:643] [c10d] The server socket on [localhost]:29501 is not yet listening (errno: 111 - Connection refused), will retry.
[I socket.cpp:582] [c10d] The client socket is attempting to connect to [localhost]:29501.
[I socket.cpp:643] [c10d] The server socket on [localhost]:29501 is not yet listening (errno: 111 - Connection refused), will retry.
[I socket.cpp:582] [c10d] The client socket is attempting to connect to [localhost]:29501.
[I socket.cpp:643] [c10d] The server socket on [localhost]:29501 is not yet listening (errno: 111 - Connection refused), will retry.
...
[W socket.cpp:598] [c10d] No socket on (127.0.0.1, 29501) is listening yet, will retry.
...
[E socket.cpp:726] [c10d] The client socket has timed out after 300s while trying to connect to (127.0.0.1, 29501).
```

Differential Revision: [D34365217](https://our.internmc.facebook.com/intern/diff/D34365217/)

### CODE REVIEW

**Comment by cbalioglu** [comment]
cc @stas00

**Comment by stas00** [comment]
Thank you for working on this, @cbalioglu 

so each process is writing this? So I still get hundreds of those even if it writes every 20 seconds.

> Instead of outputting it after every failed attempt (meaning every second), it is now written every 20 seconds. 

Do you mean it'll:

A. emit a warning immediately and then every 20 secs
B. emit nothing until 20 secs pass and then continue every 20 secs.

I vote for B, since IMHO the ideal approach is to not log anything at all until 20 secs passed giving it a warm up period. i.e. wait silently for 20 secs and only then warn that something is wrong? 99% of the time since I switched to pt-1.11 I get this noise for a few secs and then it's gone. And with 256-512 processes this is a lot of noise.

And ideally we want the fix in 1.11 if it's not too late, since this is a regression from 1.10.

Thank you!

---------------------

Could you please share why this warning / info wasn't appearing until pt-1.11? if I switch to pt-1.10 they are not there.

> With log level set to INFO (a.k.a. verbose or debug level) the output looks like:

Why is INFO = debug? I thought that's what DEBUG level was for, no?

**Comment by cbalioglu** [comment]
> Thank you for working on this, @cbalioglu

You are welcome!

> Do you mean it'll:
> 
> A. emit a warning immediately upon encountering the issue and then every 20 secs B. emit nothing until 20 secs passed and then emit one every 20 secs.

With the latest revision it is B.

> And ideally we want the fix in 1.11 if it's not too late, since this is a regression from 1.10.


> Could you please share why this warning / info wasn't appearing until pt-1.11? if I switch to pt-1.10 they are not there.
Those log statement were simply not there before v1.11. The new socket implementation was introduced after v1.10. So this will be the first official release including those changes.

> Why is INFO == debug? I thought that's what DEBUG level was for, no?

If you check the latest revision of this PR stack, you will see that I actually made quite a few changes. The C++ logging facility of PyTorch has only three log levels similar to GLOG: Info, Warning, and Error. Technically there are also VLOG macros for verbose output, but as of today there is no way to change the verbosity level. In the latest revision I leveraged our existing `TORCH_DISTRIBUTED_DEBUG` environment variable to introduce two new logging macros `C10D_DEBUG` and `C10D_TRACE`. I also revised the log statements in the socket implementation and moved most noisy ones to Debug or Trace depending on their severity.

Having said that I will revise #71746 tomorrow and will try to expedite the review process. There is still a small window for us to cherry-pick these changes for the v1.11 release.

**Comment by stas00** [comment]
Super! Thank you for explaining and improving this domain of pytorch, @cbalioglu!

Let's hope this PR can make it into pt-1.11. I'd be happy to test once there is a new RC package to try.

File: torch/csrc/distributed/c10d/socket.cpp [hunk 1]
```diff
@@ -650,20 +653,26 @@ bool SocketConnectOp::tryConnect(int family) {
         return true;
       }
 
-      if (cr == ConnectResult::TimeOut) {
-        auto msg = fmt::format(
-            "The client socket has timed out after {} while trying to connect to ({}, {}).",
-            opts_->connect_timeout(),
-            host_,
-            port_);
+      if (cr == ConnectResult::Retry) {
```
Comment (rohan-varma) [inline torch/csrc/distributed/c10d/socket.cpp:656]: nit: can we merge this into L661?

File: torch/csrc/distributed/c10d/socket.cpp [hunk 2]
```diff
@@ -650,20 +653,26 @@ bool SocketConnectOp::tryConnect(int family) {
...
+        if (retry_attempt == 30) {
+          C10D_INFO("No socket on ({}, {}) is listening yet, will retry.", host_, port_);
 
-        C10D_ERROR(msg);
+          retry_attempt = 0;
+        }
 
-        throw TimeoutError{msg};
-      }
+        // Wait one second to avoid choking the server.
```
Comment (rohan-varma) [inline torch/csrc/distributed/c10d/socket.cpp:670]: nit: wait `delay_duration_`?

**Review by rohan-varma — APPROVED** [review]
Looks great, thanks for quickly resolving @stas00 's feedback! Should be good to land and get into PT-1.11 as a cherry pick if all of @stas00 's concerns are addressed.

File: torch/csrc/distributed/c10d/socket.cpp (hunk 1 above)
Comment (cbalioglu) [reply torch/csrc/distributed/c10d/socket.cpp:656]: This is inside the `do`/`while` loop, so at this point we do not know whether another IP address in our stack will succeed before we retry from start. This keeps record whether we should retry it after we exhaust all our options.

**Pull request closed without merging**